test/engine/null.test.lua

```diff
@@ -612,3 +612,117 @@ s:insert{1, -1, 1}
+-- gh-4480: Introduce exclude_null option
+--
+
+-- Basic functionality
+s = box.schema.space.create('test', {engine=engine})
+pk = s:create_index('primary', {parts={1, 'number'}})
+sk1 = s:create_index('sk1', {parts={{2, 'number', is_nullable=true, exclude_null=true}}})
+sk1.parts
+sk2 = s:create_index('sk2', {parts={{2, 'number', is_nullable=true, exclude_null=false}}})
+sk2.parts
+
+s:insert{1, 1}
+s:insert{2, box.NULL}
+pk:select{}
+sk1:select{} -- [1, 1] only
+sk2:select{}
+
+-- Error
+sk3 = s:create_index('sk3', {parts={{2, 'number', is_nullable=false}}})
+s:truncate()
+
+-- Ok
+sk3 = s:create_index('sk3', {parts={{2, 'number', is_nullable=false}}})
+
+s:insert{1, 1}
+s:insert{2, box.NULL} -- error
+
+-- if exclude_null=true, then is_nullable=true (explicitly or implicitly)
+s:drop()
+s = box.schema.space.create('test', {engine=engine})
+pk = s:create_index('primary', {parts={1, 'number'}})
+sk1 = s:create_index('sk1', {parts={{2, 'number', is_nullable=false, exclude_null=true}}})
+sk1 = s:create_index('sk1', {parts={{2, 'number', exclude_null=true}}})
+
+-- Multipart
+s:drop()
+s = box.schema.space.create('test', {engine=engine})
+pk = s:create_index('primary', {parts={1, 'number'}})
+parts = {}
+parts[1] = {2, 'number', exclude_null=true}
+parts[2] = {3, 'number', is_nullable=true}
+sk1 = s:create_index('sk1', {parts=parts})
+
+s:insert{1, 1, 1}
+s:insert{2, box.NULL, 2}
+s:insert{3, 3, box.NULL}
+s:select{}
+sk1:select{}
+
+-- Insert, then create index
+s:drop()
+s = box.schema.space.create('test', {engine=engine})
+pk = s:create_index('primary', {parts={1, 'number'}})
+
+s:insert{1, 11}
+s:insert{2, 22}
+s:insert{3, box.NULL}
+s:insert{4, box.NULL}
+
+sk1 = s:create_index('sk1', {parts={{2, 'number', exclude_null=true}}})
+sk1.parts
+s:select{}
+sk1:select{}
+
+-- Alter to exclude_null=false
+sk1:alter({parts={{2, 'number', is_nullable=true, exclude_null=false}}})
+sk1.parts
+sk1:select{}
+
+-- Alter back to exclude_null=true
+sk1:alter({parts={{2, 'number', exclude_null=true}}})
+sk1.parts
+sk1:select{}
+
+-- Alter to the wrong format
+sk1:alter({parts={{2, 'number', is_nullable=false, exclude_null=true}}})
+sk1.parts
+
+-- Check that delete doesn't crash on sk1
+s:insert{5, box.NULL}
+sk1:select{}
+_ = s:delete{5}
+sk1:select{}
+s:select{}
+
+-- Update and upsert correctness
+_ = s:update(1, {{'=', 2, box.NULL}})
+_ = s:update(3, {{'=', 2, 33}})
+s:upsert({2, 22}, {{'=', 2, box.NULL}})
+s:upsert({5, 55}, {{'=', 2, box.NULL}})
+sk1:select{}
+s:select{}
+
+-- Replace correctness
+_ = s:replace({1, 11})
+_ = s:replace({3, box.NULL})
+sk1:select{}
+s:select{}
+
+-- Tuple without the field
+s:drop()
+s = box.schema.space.create('test')
+pk = s:create_index('pk', {type='tree', parts={{1, 'uint'}}})
+sk = s:create_index('sk', {type='tree', parts={{2, 'uint', exclude_null=true}}})
+s:replace{1}
+sk:select{}
+s:select{}
+
+-- Forbid creating primary key with exclude_null option
+s:drop()
+s = box.schema.space.create('test', {engine=engine})
```

Comment:
Forgot to drop s..
Also I would check that after restoring from snapshot exclude_null still works.

Reply:
Thx, good point! I added test, found a bug with memtx engine, so now yet another check is added (see [memtx_tree_index_build_next](https://github.com/tarantool/tarantool/pull/5481/files#diff-fb2f7ff1cd0f93acca771e67cfce349c6a4ebc4b1a602d59667aa1d0cc216a8fR1298)). Initially I thought in memtx every change should be passed through `index_replace` function.